### ISSUE TITLE
instanceof and is lose type when reassigning

### ISSUE THREAD

**Issue by Diego1149** [microsoft/TypeScript#11595]
**TypeScript Version:**  1.8.0

**Code**

``` ts
class A {
    type: number;

    constructor(t: number) {
        this.type = t;
    }

    static isB(details: A): details is B {
        return details.type === 0;
    }

    static isC(details: A): details is C {
        return details.type === 1;
    }
}

class B extends A {
    constructor() {
        super(0);
    }
    doSomething() {
        alert("doing something");
    }
}

class C extends A {
    constructor() {
        super(1);
    }
    doOtherThing() {
        alert("doing other thing");
    }
}

class D extends B {
    anotherMethod() {

    }
}

let t: A = new D();

if (A.isB(t)) {
    //t = new B(); //should work, gives compilation error
    //t = new D(); //should work, gives compilation error
    //t = new C(); //should give compilation error (it already does this)
    t.doSomething();
}
```

**Steps to Reproduce:**

1.- remove commented line t=new B(); (now it won't compile because t became of type A)
2.- remove commented line t=new D(); (now it won't compile because t became of type A) 

**Expected behavior:**
It should compile properly since t is still of type B.
This should also work for more complex cases such as
class D extends B, and you set t to an object of type D, t should retain its type of B

**Actual behavior:** 
It doesn't compile, it says t.doSomething() doesn't exist on type A.

**Comment by RyanCavanaugh** [comment]
This has been fixed
